buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy

```diff
@@ -884,6 +884,22 @@ class BuildPlugin implements Plugin<Project> {
                 // TODO: remove this once cname is prepended to transport.publish_address by default in 8.0
                 test.systemProperty 'es.transport.cname_in_publish_address', 'true'
 
+                // Set netty system properties
+                test.systemProperty('io.netty.noUnsafe', 'true')
+                test.systemProperty('io.netty.noKeySetOptimization', 'true')
+                test.systemProperty('io.netty.recycler.maxCapacityPerThread', '0')
+                def seed = Long.parseUnsignedLong(((String) project.property('testSeed')).tokenize(':').get(0), 16)
+                if (seed % 4 == 0) {
+                    // Occasionally use the unpooled allocator as we ergonomically choose it on small heaps
+                    test.systemProperty('io.netty.allocator.type', 'unpooled')
```

Comment:
We used to do this kind of randomization for es tests, but I'm not sure we need it? If we want to have a test case that randomizes this, then let's do that, but having this randomized for *all tests* means we could get unexplainable failures in tests completely unrelated to netty. Could we have netty specific tests where we do this? Perhaps it would be better to have tests specifically around small heaps, wher we would use this allocator type?

Reply:
@rjernst I think the `unpooled` path here isn't so important but the difference between setting 'io.netty.allocator.numDirectArenas` or not is. That setting triggers very different code paths now (whether to use our custom `CopyingBytesChannel` or not).
Admittedly this isn't the best argument here because you could always argue that we need better networking tests then  but: 
We only saw some initial bugs in `CopyingBytesChannel` when other tests started failing. So getting wider coverage on both code paths we have in Netty would be and was definitely it in practice right now. I don't think we'll be able to beef up the networking tests to a point where we could get comparable coverage out of them.

Reply:
How would someone completely unfamiliar with netty diagnose and debug a failure from CopyBytesSocketChannel? Randomizing these for all tests is similar to what we do in ESIntegTestCase with randomizing various distributed settings. Having a mappings test fail for reasons completely outside of the realm of expertise or the focus of the test is a bug IMO. I can live with it there because I see ESIntegTestCase as sort of distributed focused tests (testing node interaction), but doing so for all tests is extending these failures to places that are difficult to debug (eg rest tests).  

> I don't think we'll be able to beef up the networking tests to a point where we could get comparable coverage out of them.

I believe in the power of unit tests. :) Could we have randomized, targeted unit tests for CopyBytesSocketChannel?

Reply:
> I believe in the power of unit tests. :) Could we have randomized, targeted unit tests for CopyBytesSocketChannel?

In general so do I :) But in case of networking it's tricky to keep the faith. In the example I mentioned above, issues were triggered by larger messages under load at a low rate. It's hard to unit-test that kind of scenario a-priory (Tim not added a test for it after the fact) and I still prefer some ITs breaking right away to only seeing this kind of thing in the benchmarks failing (at best) or production (at worst).

> How would someone completely unfamiliar with netty diagnose and debug a failure from CopyBytesSocketChannel

Maybe they can't but the same issue holds true for a number of tests (especially in X-pack SSL) where we randomize between Netty and NIO networking already anyway. I don't think this change really makes the situation that much worse. As a matter of fact, it actually is a bit of a help in tracking down issues because once might be able to narrow things down quickly to one or the other channel implementation used.

> Having a mappings test fail for reasons completely outside of the realm of expertise or the focus of the test is a bug IMO.

I agree but this change doesn't really affect the scenario much I think. In the end it only affects the Netty tests itself and X-Pack tests for SSL (and ML) that use the Netty codebase. The bulk of `ESIntegTestCase`s use the mock nio networking anyway that isn't touched by these changes.

Reply:
>  In the end it only affects the Netty tests itself and X-Pack tests for SSL (and ML) that use the Netty codebase

This change affects all integ tests since they use real elasticsearch nodes which will pick up these randomized settings. I only meant the mappings example as something unrelated to the randomization of distributed settings, yet susceptible to random failures.

> In the example I mentioned above, issues were triggered by larger messages under load at a low rate.

Why isn't this something we can mimic in unit tests? Not just this particular pattern, but in general randomizing several axes like message size and rate.

Reply:
> This change affects all integ tests since they use real elasticsearch nodes which will pick up these randomized settings.

Right sorry, I casually referred to all the tests using real nodes as REST ... my bad :) Still we do randomize between Netty and our own NIO networking in these already anyway aren't we? I would still argue that it's generally pretty easy to pinpoint issues to networking (as they generally show up as serialization issues) and has been in the past and it's free coverage. But I see how it's a slippery slope :)

> Why isn't this something we can mimic in unit tests? Not just this particular pattern, but in general randomizing several axes like message size and rate.

We can but what most bugs we had so far came down to was running into random NIO spots like the one we have here where there's either a partial read or partial write. I think we could definitely build a class of tests for this by simulating this stuff by turning off reading from sockets temporarily so buffers fill up and then check behavior.

-> I'm not saying were shouldn't or can't move to stronger testing of the IO corner cases I'm all for it. 
But until we have that, doing what this PR suggests is better than not doing it because without it we simply don't test one possible implementation of the low-level channels used by Netty which may lead to nasty surprises?

Reply:
> Still we do randomize between Netty and our own NIO networking in these already anyway aren't we?

AFAIK that randomization is for ESIntegTestCase. We use real jvm.options that is included with the Elasticsearch distribution when running REST integ tests.

>  I would still argue that it's generally pretty easy to pinpoint issues to networking

I'm not sure that is the case. I would imagine they can be difficult to distinguish from eg a mucked up bwc version logic? And given most peopling doing test triage look quickly at the test, we have got lots of issues assigned to core/infra because they are "the cluster didn't come up" when in fact these most of the time have nothing to do with the build.

> But until we have that, doing what this PR suggests is better than not doing it because without it we simply don't test one possible implementation of the low-level channels 

That's a similar slippery slope. Adding this now lowers the incentive to improve the tests.

Reply:
@rjernst @mark-vieira ok you win. Let's see what others think then, but I'm fine giving up here and starting work on some unit-test like coverage across supported Netty configurations then.

Reply:
@mark-vieira sure we can sync up :) 

As an easy solution though: I wonder if we couldn't just as an alternative to the randomisation here simply run all the Netty tests twice (that should be really cheap ... it's ~2min on my machine) with both channel implementations. (and maybe do the same for the security tests that use Netty ... that should also only be an added 5-10 minutes).
That doesn't give us double coverage on REST and integration tests tests but at least we have both implementations covered by the existing unit tests.